### ISSUE TITLE
Test_openjdk12_j9_extended.system_s390x_linux MathLoadTest JIT crash

### ISSUE THREAD

**Issue by pshipton** [eclipse/openj9#6248]
https://ci.eclipse.org/openj9/job/Test_openjdk12_j9_extended.system_s390x_linux_Nightly/71
MathLoadTest_all_0
```
MLT stderr Unhandled exception
MLT stderr Type=Segmentation error vmState=0x00000000
MLT stderr J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00802ac0 Signal_Code=00000001
MLT stderr Handler1=000003FF7ECCA458 Handler2=000003FF7E8A7E98 InaccessibleAddress=000358000C848000
MLT stderr gpr0=000000000000000D gpr1=00000000000002F5 gpr2=0000000074B7AD28 gpr3=0000000000000006
MLT stderr gpr4=00000000000A6800 gpr5=000000000046E3B0 gpr6=0000000074B7AD3A gpr7=0003580000000035
MLT stderr gpr8=0000000000000100 gpr9=0000000000000006 gpr10=000000000C8488C8 gpr11=000358000000000D
MLT stderr gpr12=0000000000000005 gpr13=0000000000463F00 gpr14=000003FF5EF97644 gpr15=000003FF7E37E478
MLT stderr psw=000003FF5F0BAD18 mask=0705100180000000 fpc=00b8ff00 bea=000003FF5F0BC522
MLT stderr fpr0 000000000e176298 (f: 236413600.000000, d: 1.168038e-315)
MLT stderr fpr1 377fcdf96e760ac5 (f: 1853229824.000000, d: 2.281867e-41)
MLT stderr fpr2 0000005f00000044 (f: 68.000000, d: 2.015896e-312)
MLT stderr fpr3 3faca77511ce8399 (f: 298746784.000000, d: 5.596510e-02)
MLT stderr fpr4 42c0000000000000 (f: 0.000000, d: 3.518437e+13)
MLT stderr fpr5 3d7b182886ac0ce1 (f: 2259422464.000000, d: 1.540137e-12)
MLT stderr fpr6 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr fpr7 3c00000000000000 (f: 0.000000, d: 1.084202e-19)
MLT stderr fpr8 40734413509f79fe (f: 1352628736.000000, d: 3.082547e+02)
MLT stderr fpr9 3fc1640c43cb4edb (f: 1137397504.000000, d: 1.358657e-01)
MLT stderr fpr10 000003ff7e340000 (f: 2117337088.000000, d: 2.171848e-311)
MLT stderr fpr11 000003ffd0d7f153 (f: 3503812864.000000, d: 2.172533e-311)
MLT stderr fpr12 00000000000f35f8 (f: 996856.000000, d: 4.925123e-318)
MLT stderr fpr13 000003fff64fcc48 (f: 4132424704.000000, d: 2.172843e-311)
MLT stderr fpr14 0000000000000022 (f: 34.000000, d: 1.679823e-322)
MLT stderr fpr15 000003ffd5d7de78 (f: 3587694080.000000, d: 2.172574e-311)
MLT stderr 
MLT stderr Compiled_method=java/math/BigInteger.<init>([CII)V
MLT stderr Target=2_90_20190623_72 (Linux 4.4.0-141-generic)
MLT stderr CPU=s390x (4 logical CPUs) (0x1e9a00000 RAM)
MLT stderr ----------- Stack Backtrace -----------
MLT stderr (0x000003FF5F0BAD18 [<unknown>+0x0])
```

**Comment by pshipton** [comment]
https://ci.eclipse.org/openj9/job/Test_openjdk11_j9_sanity.system_s390x_linux_Nightly/71
MathLoadTest_bigdecimal_ConcurrentScavenge_0
```
MLT stderr Unhandled exception
MLT stderr Type=Segmentation error vmState=0x00000000
MLT stderr J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=ba0a8068 Signal_Code=00000001
MLT stderr Handler1=000003FF9AB4A458 Handler2=000003FF9A727E98 InaccessibleAddress=000350000C847000
MLT stderr gpr0=0000000000000012 gpr1=0000000000000025 gpr2=0000000085225F98 gpr3=0000000000000000
MLT stderr gpr4=00000000000ADE00 gpr5=0000000000475C80 gpr6=0000000085225FB4 gpr7=0003500000000030
MLT stderr gpr8=0000000000000100 gpr9=0000000000000000 gpr10=000000000C8479E0 gpr11=0003500000000012
MLT stderr gpr12=000000000000000A gpr13=0000000000462A00 gpr14=000003FF7AFFF16C gpr15=000003FF9A07DD10
MLT stderr psw=000003FF7ADFB356 mask=0705100180000000 fpc=0038ff00 bea=000003FF7ADFB184
MLT stderr fpr0 000000000000babe (f: 47806.000000, d: 2.361930e-319)
MLT stderr fpr1 482026003516d2ea (f: 890688256.000000, d: 2.747515e+39)
MLT stderr fpr2 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr fpr3 3fb5913de48f15e9 (f: 3834582528.000000, d: 8.424746e-02)
MLT stderr fpr4 4820260000000000 (f: 0.000000, d: 2.747514e+39)
MLT stderr fpr5 3d6b1211586d425b (f: 1483555456.000000, d: 7.693921e-13)
MLT stderr fpr6 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr fpr7 3c5d000000000000 (f: 0.000000, d: 6.288373e-18)
MLT stderr fpr8 4072b394bfc2cdeb (f: 3217214976.000000, d: 2.992238e+02)
MLT stderr fpr9 000003ff9436b2a0 (f: 2486612736.000000, d: 2.172030e-311)
MLT stderr fpr10 000003ff9a040000 (f: 2583953408.000000, d: 2.172078e-311)
MLT stderr fpr11 000003ffec17f16d (f: 3960992000.000000, d: 2.172759e-311)
MLT stderr fpr12 00000000000f35c8 (f: 996808.000000, d: 4.924886e-318)
MLT stderr fpr13 000003fffa2fd5c8 (f: 4197438976.000000, d: 2.172876e-311)
MLT stderr fpr14 0000000000000022 (f: 34.000000, d: 1.679823e-322)
MLT stderr fpr15 000003fff637edb8 (f: 4130860544.000000, d: 2.172843e-311)
MLT stderr 
MLT stderr Compiled_method=java/math/BigInteger.<init>([CII)V
MLT stderr Target=2_90_20190623_72 (Linux 4.4.0-128-generic)
MLT stderr CPU=s390x (4 logical CPUs) (0x1e9a1c000 RAM)
MLT stderr ----------- Stack Backtrace -----------
MLT stderr (0x000003FF7ADFB356 [<unknown>+0x0])
```

**Comment by pshipton** [comment]
Changes since previous build

https://github.com/eclipse/openj9/compare/5e4a26d...fde4c80

@fjeremic

**Comment by fjeremic** [comment]
I don't see anything obvious in the diff. Could also be some OMR change as well. I'll try to take a look at this by EOD.

**Comment by pshipton** [comment]
> Could also be some OMR change as well

OMR didn't change between the builds. Could be a latent intermittent issue then, and we got unlucky to get two failures in the same build.

**Comment by fjeremic** [comment]
Tried to reproduce this locally on my z14 Docker container but I'm unable to see a reproduction. It seems like an intermittent problem. I'll have to work off the Grinders instead and take a look at the core file to see if we can spot what the problem is.

**Comment by 0xdaryl** [comment]
@R-Santhir : perhaps this is a good non-x86 test for the parallel grinders feature.  I'm sure it will save Filip some time.

**Comment by pshipton** [comment]
Another one
https://ci.eclipse.org/openj9/job/Test_openjdk12_j9_sanity.system_s390x_linux_Nightly/72

This has to be caused by a recent change.

**Comment by fjeremic** [comment]
> @R-Santhir : perhaps this is a good non-x86 test for the parallel grinders feature. I'm sure it will save Filip some time.

@R-Santhir I can definitely be a tester for this feature. Catch me on Slack whenever you can to explain how to run parallel and we can give it a shot. I'm going to work on this using existing Grinders in the mean time.

**Comment by fjeremic** [comment]
Internal grinder shows 0/20 failure rate. External grinders [1] [2] show 1/30 failure rate. Totaling up I'm observing a 1/50 failure rate for this issue.

Looking at the core file from the most recent failure referenced in https://github.com/eclipse/openj9/issues/6248#issuecomment-505411451 the method in question (`java/math/BigInteger.<init>([CII)V`) is a hot method, which may explain the low reproducibility rate. Why we crash is not very clear as kca is having trouble parsing the core files. Regardless, we'll need a log.

One interesting question is that the javacore reports:

```
1XHEXCPMODULE  Inside compiled method: java/math/BigInteger.<init>([CII)V
```

However the method it reports is not found on the backtrace of any thread! How can that be? @pshipton any ideas?

[1] https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/415
[2] https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/416

**Comment by fjeremic** [comment]
Launched 3x20 grinders to try and collect a log:

https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/419/console
https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/420/console
https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/421/console

**Comment by pshipton** [comment]
> However the method it reports is not found on the backtrace of any thread! How can that be?

I thought methods got queued for compilation, so execution of the method could be finished by the time the compiler does the compilation.

**Comment by fjeremic** [comment]
> I thought methods got queued for async compilation, so execution of the method could be finished by the time the compiler does the compilation.

This isn't a crash in a compilation thread. It's a crash in a JIT compiled method body.

**Comment by pshipton** [comment]
Seems weird then, indicating another bug somewhere. Could the method name being printed be incorrect?

**Comment by fjeremic** [comment]
> Could the method name being printed be incorrect?

It would seem very unlikely given the PSW register indeed does point to `java/math/BigInteger.<init>([CII)V`

**Comment by fjeremic** [comment]
Looking at the failure from [1] the crash point is at:

```
Unhandled exception
Type=Segmentation error vmState=0x00000000
J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
Handler1=000003FF9E94A430 Handler2=000003FF9E527E98 InaccessibleAddress=000358000C851000
gpr0=000000000000000C gpr1=0000000000000007 gpr2=000000008512E158 gpr3=000000000000000C
gpr4=00000000000A6800 gpr5=0000000000466390 gpr6=000000008512E168 gpr7=0003580000000036
gpr8=0000000000000100 gpr9=0000000000000007 gpr10=000000000C851B68 gpr11=000358000000000C
gpr12=0000000000000004 gpr13=0000000000460900 gpr14=000003FF7ED85E24 gpr15=000003FF9DEFE478
psw=000003FF7EBA104E mask=0705100180000000 fpc=0038ff00 bea=000003FF7EBA0D82
```

Taking a look at the log:

```
     0x3ff7eba104e 0000085e [     0x3ff1a2eafd0] e3 07 a0 08 00 76          LB      GPR0, Shadow[<array-shadow>] 8(GPR7,GPR10)
     0x3ff7eba1054 00000864 [     0x3ff1a2eb0c0] 12 00                      LTR     GPR0,GPR0
     0x3ff7eba1056 00000866 [     0x3ff1a2eb880]                            ASSOCREGS
```

We tried to load a byte from that location. `GPR7` seems to contain some garbage in the high word. From what I can tell `GPR7` should be a 64-bit register. Now looking at the trees:

```
------------------------------
 n2929n   (  0)  ificmplt --> block_1286 BBStart at n13709n ()                                        [     0x3ff35fb3340] bci=[23,7,211] rc=0 vc=18 vn=- li=- udi=- nc=3 flg=0x20
 n2923n   (  2)    b2i (in GPR_1554) (unneededConv )                                                  [     0x3ff35fb3160] bci=[23,4,210] rc=2 vc=18 vn=- li=- udi=- nc=1 flg=0x400
 n2922n   (  0)      bloadi  <array-shadow>[#235  Shadow] [flags 0x80000601 0x0 ] (in GPR_1554) (cannotOverflow signExtendedTo32BitAtSource )  [     0x3ff35fb3110] bci=[23,4,210] rc=0 vc=18 v n2921n   (  0)        aladd (X>=0 internalPtr sharedMemory )                                         [     0x3ff35fb30c0] bci=[23,4,210] rc=0 vc=228 vn=- li=- udi=- nc=2 flg=0x8100
 n14616n  (  3)          ==>aRegLoad (in &GPR_1540) (X!=0 SeenRealReference sharedMemory )
 n2920n   (  0)          lsub (in GPR_1545) (highWordZero X>=0 cannotOverflow )                       [     0x3ff35fb3070] bci=[23,4,210] rc=0 vc=18 vn=- li=- udi=- nc=2 flg=0x5100
 n2919n   (  0)            i2l (in GPR_1545) (highWordZero X>=0 )                                     [     0x3ff35fb3020] bci=[23,4,210] rc=0 vc=18 vn=- li=- udi=- nc=1 flg=0x4100
 n4470n   (  0)              ==>su2i (in GPR_1545) (X>=0 unneededConv )
 n2918n   (  0)            lconst -8 (X!=0 X<=0 )                                                     [     0x3ff35fb2fd0] bci=[23,4,210] rc=0 vc=18 vn=- li=- udi=- nc=0 flg=0x204
 n2926n   (  0)    iconst 0 (X==0 X>=0 X<=0 )                                                         [     0x3ff35fb3250] bci=[23,7,211] rc=0 vc=18 vn=- li=- udi=- nc=0 flg=0x302
 n14638n  (  0)    GlRegDeps ()                                                                       [     0x3ff199bcf30] bci=[23,7,211] rc=0 vc=18 vn=- li=- udi=- nc=3 flg=0x20
 n14612n  (  3)      ==>aRegLoad (in &GPR_1536)
 n14614n  (  2)      ==>iRegLoad (in GPR_1538) (X>=0 cannotOverflow SeenRealReference )
 n14616n  (  3)      ==>aRegLoad (in &GPR_1540) (X!=0 SeenRealReference sharedMemory )
------------------------------

 [     0x3ff1a2eafd0]   LB      GPR_1554, Shadow[<array-shadow>] 8(GPR_1545,&GPR_1540)
 [     0x3ff1a2eb0c0]   LTR     GPR_1554,GPR_1554
 [     0x3ff1a2eb880]   ASSOCREGS
  PRE:
 {GPR2:&GPR_1536:R} {GPR8:GPR_1538:R} {GPR10:&GPR_1540:R}
 [     0x3ff1a2eb2b0]   BRC     BM(0x4), Label L0076
 POST:
 {GPR2:&GPR_1536:R} {GPR8:GPR_1538:R} {GPR10:&GPR_1540:R}
```

It seems we have skipped the evaluation of the `i2l` node at `n2919n` and assigned it virtual register `GPR_1545` (the child). The node is not marked as `unneededConv`, and rightfully so because we do need to perform sign extension here because we are about to use the node within a memory reference. This seems like the reason why we crash. I'm looking into why we skipped evaluation of this sign extension.

[1] https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/421/consoleFull

**Comment by fjeremic** [comment]
Just dropping a quick comment here that I've identified the bug in the code and am now working towards the fix. I'll detail what exactly is wrong in the PR. We should hopefully have a fix tomorrow.

**Comment by fjeremic** [comment]
Grinders with the fix passed 160/160:

https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/432/
https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/433/
https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/434/
https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/435/
https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/436/
https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/437/
https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/438/
https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/439/

Given the reported failure rate of 1/30 I'm confident the fix has worked. @0xdaryl FYI would appreciate a review of the OMR PR.

**Comment by pshipton** [comment]
https://ci.eclipse.org/openj9/job/Test_openjdk12_j9_extended.system_s390x_linux_Nightly/85

**Comment by fjeremic** [comment]
Fix is still there in eclipse/omr#4103 waiting to be reviewed and merged. @0xdaryl could you be a committer for this issue given that it's popping up more frequently now?

**Comment by pshipton** [comment]
https://ci.eclipse.org/openj9/job/Test_openjdk12_j9_sanity.system_s390x_linux_Nightly/87
